### PR TITLE
etcd: update version to 3.1.10

### PR DESCRIPTION
**Pull request by hongchaodeng**
ref: https://github.com/kubernetes/kubernetes/issues/49386

Need image pushed:
```
gcr.io/google_containers/etcd:3.1.10
```

### CODE REVIEW

**Comment by luxas** [comment]
@hongchaodeng Thanks!

cc @wojtek-t on pushing the images

**Comment by luxas** [comment]
cc @kubernetes/sig-scalability-pr-reviews @kubernetes/sig-cluster-lifecycle-pr-reviews

**Comment by timothysc** [comment]
@hongchaodeng I thought you we're waiting on a fix.?.?

**Comment by hongchaodeng** [comment]
@timothysc 
The fix is self-hosted related problem. We can update the version while fixing the bug. It's parallel.

Client library will be updated too. It's a little tricky due to grpc intertwining between all deps.

**Comment by timothysc** [comment]
> Client library will be updated too. It's a little tricky due to grpc intertwining between all deps. 

I've done that many times now, I can take a stab at it.

**Comment by hongchaodeng** [comment]
Sure. Great that you know how to solve it ;-)

**Comment by wojtek-t** [comment]
@jpbetz @lavalamp @smarterclayton - I think we should do this, but would also like to get your opinion on it.

**Comment by edevil** [comment]
Is there some incompatibility with etcd 3.2.x? I've tried using it and kubernetes seemed to work fine.

**Comment by luxas** [comment]
@edevil Not that I know of, but seems like it's best to upgrade one minor version at a time in order to be safe. That's what's officially supported.

**Comment by jpbetz** [comment]
Looks good to me.  I've done a quick glance over the changes in 3.1.x (https://github.com/coreos/etcd/releases/tag/v3.1.0 and similar) .

For posterity, here's the 3.1 upgrade notes:

https://github.com/coreos/etcd/blob/master/Documentation/upgrades/upgrade_3_1.md

**Comment by timothysc** [comment]
@ixdy We need an image push before we can finish this PR.

**Review by timothysc — APPROVED** [review]
/lgtm @wojtek-t ?

File: build/root/WORKSPACE
Comment (mikedanese) [inline build/root/WORKSPACE]: The "sha256" value in this rule is going to change. I can't comment on the line but it's line 20. The new hash is "2d335f298619c6fb02b1124773a56966e448ad9952b26fea52909da4fe80d2be" according to the build failure.

File: cluster/images/etcd/Makefile
Comment (ixdy) [inline cluster/images/etcd/Makefile]: do we really need to include all 4 etcd versions in the 3.1.10 image?

Comment (hongchaodeng) [inline cluster/images/etcd/Makefile]: What do you want to change to?

Comment (timothysc) [inline cluster/images/etcd/Makefile]: Depends on supported versions of kube doesn't it.  What's the oldest supported version of kube?

Comment (wojtek-t) [inline cluster/images/etcd/Makefile]: If we want to support 2.2 -> 3.1 upgrade with this script, then all versions are neceesary. And we still officially support etcd v2 (which means probably 2.2.1 because we never really qualified 2.3.*).

Comment (ixdy) [inline cluster/images/etcd/Makefile]: I was not sure whether we still needed to support etcd 2.2, but it sounds like we do, so OK.

**Comment by jagosan** [comment]
@mml - this issue stalled out a bit and needs a kick. Adding you to increase odds of making progress...

**Comment by mml** [comment]
I think there needs to be a features repo issue for this?  Because there is a lot more to do if we're pushing this in 1.8, and we'll have to back it out if we don't get it all done.

1. Do we de-support v2 after this?  Or at the very least deprecate it?
2. Do we have upgrade and rollback test coverage yet?

**Comment by jpbetz** [comment]
@hongchaodeng let us know when you're for images to be published to `gcr.io/google_containers/etcd`. A number of us at Google (@mml, @lavalamp, @wojtek-t, etc..) have permissions to publish. I've also requested permissions.

**Comment by wojtek-t** [comment]
@jpbetz - before creating the images, I would like to have https://github.com/kubernetes/kubernetes/pull/49629 to be merged. Once this is done, I think we can build the image.

**Comment by jpbetz** [comment]
@hongchaodeng What's the status of this? #49629 has been merged. 

Should I publish etcd images? I have all necessary privileges.
Are we still targeting k8s 1.8? If so we need to get this added to the 1.8 release notes.

**Comment by dims** [comment]
@jpbetz sounds risky given the the code freeze tomorrow. right?

**Comment by hongchaodeng** [comment]
> Are we still targeting k8s 1.8?

No targets.

> Should I publish etcd images?

Yes.

**Comment by hongchaodeng** [comment]
@mml 

> Do we de-support v2 after this? Or at the very least deprecate it?

Re. compatibility, please see:
https://github.com/coreos/etcd/blob/master/Documentation/upgrades/upgrade_3_1.md .

> Do we have upgrade and rollback test coverage yet?

What's the context? This should be installer tool specific. Please discuss it in cluster lifecycle sig.

**Comment by timothysc** [comment]
Hey folks, what's going on with this one?  It's lingered and the client has already been updated in the code.

**Comment by hongchaodeng** [comment]
Still waiting the image to be pushed.

Then we will need all approvers to approve:

https://github.com/kubernetes/kubernetes/pull/49393#issuecomment-317775338

**Comment by ericchiang** [comment]
> Still waiting the image to be pushed.

@hongchaodeng who owns that?

**Comment by luxas** [comment]
A googler

**Comment by jpbetz** [comment]
I'll publish momentarily.

**Comment by jpbetz** [comment]
Sorry for the delay.

I've pushed these images from a checkout of this PR:

- gcr.io/google_containers/etcd:3.1.10
- gcr.io/google_containers/etcd-amd64:3.1.10
- gcr.io/google_containers/etcd-arm64:3.1.10
- gcr.io/google_containers/etcd-ppc64le:3.1.10
- gcr.io/google_containers/etcd-s390x:3.1.10

**Comment by luxas** [comment]
@jpbetz what about `gcr.io/google_containers/etcd-arm:3.1.10`?

**Comment by hongchaodeng** [comment]
./hack/test-update-storage-objects.sh is still trying to verify v2 keys:
```shell
kube::log::status "Verifying ${resource}/${namespace}${name} has storage version ${old_storage_version} in etcd"
curl -s http://${ETCD_HOST}:${ETCD_PORT}/v2/keys/${ETCD_PREFIX}/${resource}/${namespace}${name} | grep ${old_storage_version}
```

How could the test pass before ?!

**Comment by hongchaodeng** [comment]
What should be done with `./hack/test-update-storage-objects.sh` where it is still testing migration from v2 to v3? Since we are upgrading to 3.1.x, shouldn't it be removed?

**Comment by jpbetz** [comment]
@luxas, I had missed `arm`.  Publishing it now.

**Comment by hongchaodeng** [comment]
fixed `./hack/test-update-storage-objects.sh`

**Comment by luxas** [comment]
Thanks @jpbetz!

**Comment by wojtek-t** [comment]
It's too late to have it for 1.8.
We should merge this as soon as master is open for 1.9 PRs.

**Comment by luxas** [comment]
@wojtek-t Can we start running the density/kubemark-5000 suites as soon as v1.9 coding opens as well so we can make sure this will make v1.9?

**Comment by luxas** [comment]
Can we merge/rebase this now to get this in early for new clusters v1.9?

File: federation/manifests/federation-apiserver-deployment.yaml
Comment (timothysc) [inline federation/manifests/federation-apiserver-deployment.yaml]: I'm confused why this change is here?

Comment (hongchaodeng) [reply federation/manifests/federation-apiserver-deployment.yaml]: Seems like it's been deleted. Mistakenly added in rebase. Will remove

**Review by timothysc — COMMENTED** [review]
/lgtm 

But someone who owns the upstream jiggery should approve.

**Review by luxas — APPROVED** [review]
/approve
for cmd/kubeadm

**Comment by hongchaodeng** [comment]
/assign ixdy

Approve for `build/OWNERS` ?

File: build/build-image/cross/Dockerfile
Comment (ixdy) [inline build/build-image/cross/Dockerfile]: need to update build/build-image/cross/VERSION as well.

Comment (hongchaodeng) [reply build/build-image/cross/Dockerfile]: Increment the last number? Like v1.8.3-2 -> v1.8.3-3 ?

Comment (luxas) [reply build/build-image/cross/Dockerfile]: yes

**Comment by wojtek-t** [comment]
@jpbetz @hongchaodeng - can you please take a look into those failures?

**Comment by hongchaodeng** [comment]
@wojtek-t 
I did take a look at some of them before. Those I looked at are transient build failures like:

```
main.go:260: Something went wrong: failed to acquire k8s binaries: error during make -C /go/src/k8s.io/kubernetes quick-release: exit status 2
```

Any suggestions? I think they are more infra problems and hope that rerunning later after infra is fixed.

**Comment by jpbetz** [comment]
From the build logs:
```
Step 1/16 : FROM gcr.io/google_containers/kube-cross:v1.8.3-3
manifest for gcr.io/google_containers/kube-cross:v1.8.3-3 not found
```

I'll try to publish `1.8.3-3` to `google_containers` now.

**Comment by hongchaodeng** [comment]
@jpbetz Thanks!
Seems my fault not digging deep enough...

**Comment by jpbetz** [comment]
```

**Comment by jpbetz** [comment]
gcr.io/google_containers/kube-cross:v1.8.3-3 published.

```
git fetch origin pull/49393/head:pr49393 # This PR
git co pr49393
git rebase master
cd build/build-image/cross
make build
make push
```

/cc @ixdy @mikedanese 
/retest

**Comment by mikedanese** [comment]
```console
$ docker push gcr.io/google_containers/kube-cross:v1.8.3-3
The push refers to a repository [gcr.io/google_containers/kube-cross]
v1.8.3-3: digest: sha256:7de0c648dfc17bd3162706b7f1fc2c6467fb5ebba722d702c1611eaf68b34a9f size: 3269
```

**Pull request merged**
